src/Passes/Source/RuleTransformationPass/Profile.cpp

```diff
@@ -442,14 +440,32 @@ namespace quantum
             function->clearGC();
             function->clearMetadata();
 
+            // Deleteing blocks in reverse order
+            std::vector<llvm::BasicBlock*> blocks;
             for (auto& block : *function)
             {
-                llvm::errs() << "REMOVING" << block << "\n";
-                // Removing all instructions
+                blocks.push_back(&block);
+            }
+
+            // Removing backwards to avoid segfault
+            for (auto it = blocks.rbegin(); it != blocks.rend(); ++it)
+            {
+                auto& block = **it;
+                // Deleting instructions in reverse order (needed bauase it is a DAG structure)
+                std::vector<llvm::Instruction*> instructions;
                 for (auto& instr : block)
                 {
+                    instructions.push_back(&instr);
+                }
+
+                // Removing all instructions
+                for (auto it2 = instructions.rbegin(); it2 != instructions.rend(); ++it2)
+                {
+                    auto& instr = **it2;
+                    llvm::errs() << " -- " << instr << "\n";
                     instr.replaceAllUsesWith(llvm::UndefValue::get(instr.getType()));
-                    to_delete.push_back(&instr);
+                    instr.eraseFromParent();
+                    // to_delete.push_back(&instr);
```

Comment:
I think this commented line can be removed now, as the `to_delete` variable is no longer used anywhere else.

Reply:
Remove the comment